src/modules/opentsdb.c

```diff
@@ -0,0 +1,424 @@
+      if (fourth_space == NULL) {
+        mtevL(nldeb, "Invalid OpenTSDB record, can't find the fourth space or tab in: %s\n", record);
+        continue;
+      }
+    }
+    *fourth_space++ = 0;
+    while (*fourth_space == ' ' || *fourth_space == '\t') fourth_space++;
+
+    mtevL(nldeb, "OpenTSDB record: %s\n", record);
+    char *opentsdb_metric_name = first_space;
+    char *opentsdb_timestamp = second_space;
+    char *opentsdb_value = third_space;
+    char *opentsdb_tags = fourth_space;
+
+    // timestamps for telnet are allowed to contain a '.' before the milliseconds
+    // otherwise, if it is 10 digits or less it is seconds, more than 10 digits is ms
+    char *dp;
+    uint64_t whence_ms = 0;
+    if (strlen(opentsdb_timestamp) <= 10)
+    {
+      whence_ms = strtoull(opentsdb_timestamp, &dp, 10);
+    }
+    else {
+      whence_ms = strtoull(opentsdb_timestamp, &dp, 10) * 1000;
+      if (dp && *dp == '.')
+        whence_ms += (int)(1000.0 * atof(dp));
+    }
+
+    size_t opentsdb_value_len = strlen(opentsdb_value);
+    if (count_integral_digits(opentsdb_value, opentsdb_value_len, mtev_true) == 0) {
+      mtevL(nldeb, "Invalid OpenTSDB record, no digits in value: %s\n", third_space);
+      continue;
+    }
+
+    double metric_value = 0.0;
+    const char *dot = strchr(opentsdb_value, '.');
+    if (dot == NULL) {
+      /* attempt fast integer parse, if we fail, jump to strtod parse */
+      /* OpenTSDB only supports signed int64 */
+      errno = 0;
+      int64_t v = strtoll(opentsdb_value, NULL, 10);
+      if (errno == ERANGE) {
+        goto strtod_parse;
+      }
+      metric_value = (double)v;
+    } else {
+    strtod_parse:
+      errno = 0;
+      metric_value = strtod(opentsdb_value, NULL);
+      if (errno == ERANGE) {
+        mtevL(nldeb, "Invalid opentsdb record, strtod cannot parse value: %s\n", third_space);
+        continue;
+      }
+    }
+
+    /* allow for any length name + tags in the broker */
+    mtev_dyn_buffer_t tagged_name;
+    mtev_dyn_buffer_init(&tagged_name);
+
+    /* OpenTSDB tags are k=v list with at least one kv, each kv separated by space
+     * Re-format incoming name string into our tag format for parsing */
+    mtev_dyn_buffer_add(&tagged_name, (uint8_t *)opentsdb_metric_name, strlen(opentsdb_metric_name));
+    mtev_dyn_buffer_add(&tagged_name, (uint8_t *)"|ST[", 4);
+    /* look for K=V pairs */
+    char *pair, *lasts;
+    bool comma = false;
+    char *space_loc = opentsdb_tags;
+    for (pair = strtok_r(space_loc, " ", &lasts); pair; pair = strtok_r(NULL, " ", &lasts)) {
+      const char *equal = strchr(pair, '=');
+      if (equal) {
+        size_t pair_len = strlen(pair);
+        if (!noit_metric_tagset_is_taggable_key(pair, equal - pair) ||
+            !noit_metric_tagset_is_taggable_value(equal + 1, pair_len - ((equal + 1) - pair))) {
+          mtevL(nldeb, "Unacceptable tag key or value: '%s' skipping\n", pair);
```

Comment:
this is out of scope because already present in existing graphite code for both reconnoiter and snowth, will make this fix in a separate PR for each.
it actually appears that we can drop this entirely, because all the `noit_stats_xxx()` functions call `build_tag_extended_name()` which calls `noit_metric_canonicalize()` and so the tags are examined lower down in this stack with `noit_metric_tagset_is_taggable_xxx()`, truncated uniquely, sorted and encoded there.